### PR TITLE
crimson/os/seastore: introduce (64-8)-bit signed device_off_t and 32-bit signed segment_off_t

### PR DESCRIPTION
**Pull request by cyx1231st**
They relaces the original 64-bit unsigned `block_off_t` and 32-bit signed `seastore_off_t` for the following considerations:
* The offset for `blk_paddr_t` should not be 32-bit `seastore_off_t`, following up https://github.com/ceph/ceph/pull/47698#discussion_r951039417
* Extend the offset type of `res_paddr_t` for the same reason.
* Fix `extent_len_t` usages for extent length and block size.
* The size/offset for both circular/segmented journal should not be limited by 32-bit `seastore_off_t`.
* Re-introduce `segment_off_t` to represent the 32-bit segment offset field from `seg_paddr_t`.

Does it address the issue you see @aravind-wdc ?

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
jenkins test api

File: src/crimson/os/seastore/seastore_types.cc [hunk 1]
```diff
@@ -132,21 +132,21 @@ std::ostream &operator<<(std::ostream &out, const paddr_t &rhs)
...
   segment_seq_t jseq = segment_seq;
-  seastore_off_t joff;
+  device_off_t joff;
   if (type == journal_type_t::SEGMENTED) {
     joff = offset.as_seg_paddr().get_segment_off();
   } else {
     assert(type == journal_type_t::RANDOM_BLOCK);
-    auto boff = offset.as_blk_paddr().get_block_off();
+    auto boff = offset.as_blk_paddr().get_device_off();
     assert(boff <= MAX_SEG_OFF);
```
Comment (myoungwon) [inline src/crimson/os/seastore/seastore_types.cc]: assert(boff <= DEVICE_OFF_MAX) ?

File: src/crimson/os/seastore/seastore_types.h [hunk 2]
```diff
@@ -700,7 +709,7 @@ struct seg_paddr_t : public paddr_t {
     internal_paddr |= static_cast<u_segment_off_t>(off);
   }
 
-  paddr_t add_offset(seastore_off_t o) const {
+  paddr_t add_offset(device_off_t o) const {
     device_off_t off = get_segment_off() + o;
```
Comment (myoungwon) [inline src/crimson/os/seastore/seastore_types.h:689]: We may need to check o is less than max value of segment_off_t?

File: src/crimson/os/seastore/seastore_types.cc (hunk 1 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/seastore_types.cc]: Dropped in the last commit. And boff from `blk_paddr_t::get_device_off()` is guaranteed to be reasonable from its constructor.

File: src/crimson/os/seastore/seastore_types.h (hunk 2 above)
Comment (cyx1231st) [reply src/crimson/os/seastore/seastore_types.h:689]: Hmm, it should be good as long as the `off` for `seg_paddr_t` is reasonable.

**Review by myoungwon — APPROVED** [review]

**Comment by aravind-wdc** [comment]
> 

@cyx1231st Yes it does solve my issue. Thanks.

File: src/crimson/os/seastore/segment_manager/ephemeral.h [hunk 3]
```diff
@@ -20,6 +20,15 @@ struct ephemeral_config_t {
...
   size_t segment_size = 0;
+
+  void validate() const {
+    ceph_assert_always(size > 0);
+    ceph_assert_always(size <= DEVICE_OFF_MAX);
+    ceph_assert_always(segment_size > 0);
+    ceph_assert_always(segment_size <= MAX_SEG_OFF);
+    ceph_assert_always(size / segment_size > 0);
+    ceph_assert_always(size / segment_size <= DEVICE_SEGMENT_ID_MAX);
+  }
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/segment_manager/ephemeral.h:31]: why do these two validates have different content?

Comment (cyx1231st) [reply src/crimson/os/seastore/segment_manager/ephemeral.h:31]: They are logically identical to me except that ephemeral uses `ephemeral_config_t` and zns uses `zns_sm_metadata_t`.

**Review by athanatos — APPROVED** [review]
Yep, makes sense to me!

**Review by liu-chunmei — APPROVED** [review]

**Pull request merged**